Source/Painting/SvgLinearGradientServer.cs

```diff
@@ -78,44 +69,37 @@ public override Brush GetBrush(SvgVisualElement renderingElement, ISvgRenderer r
                         transform.Scale(bounds.Width, bounds.Height, MatrixOrder.Prepend);
-                        transform.RotateAt(-90.0f, midPoint, MatrixOrder.Prepend);
                     }
                     transform.TransformPoints(points);
                 }
 
+                points[0].X = (float)Math.Round(points[0].X, 4);
+                points[0].Y = (float)Math.Round(points[0].Y, 4);
+                points[1].X = (float)Math.Round(points[1].X, 4);
+                points[1].Y = (float)Math.Round(points[1].Y, 4);
+
                 if (this.GradientUnits == SvgCoordinateUnits.ObjectBoundingBox)
                 {
                     // Transform the normal line back to a line such that the gradient still starts in the correct corners, but
                     // has the proper normal vector based on the transforms.  If you work out the geometry, these formulas should work.
                     var midPoint = new PointF((points[0].X + points[1].X) / 2, (points[0].Y + points[1].Y) / 2);
-                    var dy = (points[1].Y - points[0].Y);
-                    var dx = (points[1].X - points[0].X);
-                    var x2 = points[0].X;
+                    var dy = points[1].Y - points[0].Y;
+                    var dx = points[1].X - points[0].X;
+                    var x1 = points[0].X;
                     var y2 = points[1].Y;
 
-                    if (Math.Round(dx, 4) == 0)
-                    {
-                        points[0] = new PointF(midPoint.X + dy / 2 * bounds.Width / bounds.Height, midPoint.Y);
-                        points[1] = new PointF(midPoint.X - dy / 2 * bounds.Width / bounds.Height, midPoint.Y);
-                    }
-                    else if (Math.Round(dy, 4) == 0)
-                    {
-                        points[0] = new PointF(midPoint.X, midPoint.Y - dx / 2 * bounds.Height / bounds.Width);
-                        points[1] = new PointF(midPoint.X, midPoint.Y + dx / 2 * bounds.Height / bounds.Width); ;
-                    }
-                    else
+                    if (dx != 0f && dy != 0f)
                     {
-                        var startX = (float)((dy * dx * (midPoint.Y - y2) + Math.Pow(dx, 2) * midPoint.X + Math.Pow(dy, 2) * x2) /
+                        var startX = (float)((dy * dx * (midPoint.Y - y2) + Math.Pow(dx, 2) * midPoint.X + Math.Pow(dy, 2) * x1) /
                         (Math.Pow(dx, 2) + Math.Pow(dy, 2)));
-                        var startY = dy * (startX - x2) / dx + y2;
-                        points[0] = new PointF(startX, startY);
-                        points[1] = new PointF(midPoint.X + (midPoint.X - startX), midPoint.Y + (midPoint.Y - startY));
+                        var endY = dy * (startX - x1) / dx + y2;
+                        points[0] = new PointF(startX, midPoint.Y + (midPoint.Y - endY));
+                        points[1] = new PointF(midPoint.X + (midPoint.X - startX), endY);
                     }

```

Comment:
This looks correct, as far as I can see, but I don't understand why the original problem (gradients > 698px not rendered) happend. Can you explain this a bit?

Reply:
Although it is a gradation in horizontal direction, `dx` does not become zero due to error due to rotation calculation.

`dx = -0.0000610351563`

As a result, diagonal line is calculated.

In case of horizontal (or vertical), starting and ending point do not need to be converted, so processing in that case is also deleted.